homeassistant/components/mqtt/device_automation.py

```diff
@@ -0,0 +1,44 @@
+"""Provides device automations for MQTT."""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components import mqtt
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
+from . import ATTR_DISCOVERY_HASH, device_trigger
+from .discovery import MQTT_DISCOVERY_NEW, clear_discovery_hash
+
+_LOGGER = logging.getLogger(__name__)
+
+AUTOMATION_TYPE_TRIGGER = "trigger"
+AUTOMATION_TYPES = [AUTOMATION_TYPE_TRIGGER]
+AUTOMATION_TYPES_SCHEMA = vol.In(AUTOMATION_TYPES)
+CONF_AUTOMATION_TYPE = "automation_type"
+
+PLATFORM_SCHEMA = mqtt.MQTT_BASE_PLATFORM_SCHEMA.extend(
+    {vol.Required(CONF_AUTOMATION_TYPE): AUTOMATION_TYPES_SCHEMA},
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up MQTT device automation dynamically through MQTT discovery."""
+
+    async def async_discover(discovery_payload):
+        """Discover and add an MQTT device automation."""
+        try:
+            discovery_hash = discovery_payload.pop(ATTR_DISCOVERY_HASH)
+            config = PLATFORM_SCHEMA(discovery_payload)
+            if config[CONF_AUTOMATION_TYPE] == AUTOMATION_TYPE_TRIGGER:
+                await device_trigger.async_setup_trigger(
+                    hass, config, config_entry, discovery_hash
+                )
+        except Exception:
```

Comment:
Use `finally` instead. 

Also, only wrap the code in try…except that can raise.

Reply:
Not sure how to use `finally` here?

> Also, only wrap the code in try…except that can raise.

You mean to move this outside the try..except block:
`discovery_hash = discovery_payload.pop(ATTR_DISCOVERY_HASH)`
?

Reply:
You are catching all exceptions and then reraising, that is exactly like finally.

```python
try:
    await something_that_might_raise
finally:
    if X:
        clear(X)
```

Reply:
Exception handling tidied up a bit as discussed on discord.